File: leetcode-75/10-binary-tree-bfs/40-maximum-level-sum-of-a-binary-tree/solution.ts
/**
 * LeetCode 1161. Maximum Level Sum of a Binary Tree
 * https://leetcode.com/problems/maximum-level-sum-of-a-binary-tree/
 *
 * Approach: BFS level order traversal, summing each level's node
 * values and tracking the level with the maximal sum (first one wins
 * on a tie because we scan levels in increasing order and only update
 * on a strictly greater sum).
 *
 * Time:  O(n) - every node is visited once
 * Space: O(n) - queue holds up to the widest level of the tree
 */

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function maxLevelSum(root: TreeNode | null): number {
  if (root === null) return 0;

  const queue: TreeNode[] = [root];
  let currentLevel = 0;
  let bestLevel = 1;
  let bestSum = -Infinity;

  while (queue.length > 0) {
    currentLevel++;
    const levelSize = queue.length;
    let levelSum = 0;

    for (let i = 0; i < levelSize; i++) {
      const node = queue.shift() as TreeNode;
      levelSum += node.val;

      if (node.left) queue.push(node.left);
      if (node.right) queue.push(node.right);
    }

    if (levelSum > bestSum) {
      bestSum = levelSum;
      bestLevel = currentLevel;
    }
  }

  return bestLevel;
}
function assertEqual(actual: unknown, expected: unknown, label: string): void {
  const a = JSON.stringify(actual);
  const e = JSON.stringify(expected);
  if (a !== e) {
    console.error(`FAIL [${label}]: got ${a}, expected ${e}`);
  } else {
    console.log(`PASS [${label}]`);
  }
}

if (require.main === module) {
  // Example usage:
  const root = new TreeNode(1, new TreeNode(7, new TreeNode(7), new TreeNode(-8)), new TreeNode(0));
  console.log(maxLevelSum(root)); // 2

  // Edge cases
  assertEqual(maxLevelSum(new TreeNode(5)), 1, "single node");

  const allNegative = new TreeNode(-1, new TreeNode(-2), new TreeNode(-3));
  assertEqual(maxLevelSum(allNegative), 1, "all negative values, level 1 sum is least negative");

  const tie = new TreeNode(0, new TreeNode(0), new TreeNode(0));
  assertEqual(maxLevelSum(tie), 1, "tie between levels resolves to smallest level");

  const leftChain = new TreeNode(1, new TreeNode(2, new TreeNode(3, new TreeNode(4))));
  assertEqual(maxLevelSum(leftChain), 4, "pure left chain, deepest level has largest sum");
}

export { maxLevelSum, TreeNode };

File: leetcode-75/10-binary-tree-bfs/40-maximum-level-sum-of-a-binary-tree/solution.py
"""
LeetCode 1161. Maximum Level Sum of a Binary Tree
https://leetcode.com/problems/maximum-level-sum-of-a-binary-tree/

Approach: BFS level order traversal, summing each level's node
values and tracking the level with the maximal sum (first one wins
on a tie because we scan levels in increasing order and only update
on a strictly greater sum).

Time:  O(n) - every node is visited once
Space: O(n) - queue holds up to the widest level of the tree
"""

from collections import deque
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def maxLevelSum(self, root: Optional[TreeNode]) -> int:
        queue = deque([root])
        current_level = 0
        best_level = 1
        best_sum = float("-inf")

        while queue:
            current_level += 1
            level_size = len(queue)
            level_sum = 0

            for _ in range(level_size):
                node = queue.popleft()
                level_sum += node.val

                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)

            if level_sum > best_sum:
                best_sum = level_sum
                best_level = current_level

        return best_level


if __name__ == "__main__":
    sol = Solution()
    root = TreeNode(1, TreeNode(7, TreeNode(7), TreeNode(-8)), TreeNode(0))
    print(sol.maxLevelSum(root))  # 2

    root2 = TreeNode(989, None, TreeNode(10250, TreeNode(98693, None, TreeNode(-32127)), TreeNode(-89388)))
    print(sol.maxLevelSum(root2))  # 2
